airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksStorageOperations.kt

```diff
@@ -0,0 +1,172 @@
+        // The delete operation has to be done on a file level. You can NOT delete a folder and its
+        // contents
+        // recursively.
+        if (purgeStagedFiles) {
+            log.info {
+                "Removing staged file ${stagingDirectory(streamConfig.id, database)}/${buffer.filename}"
+            }
+            // Using Jdbc for PUT 'file' and REMOVE 'file' just returns a presigned S3 url and the
+            // HTTP method to call as
+            // in sql row data. Using workspace client instead.
+            // destinationHandler.execute(Sql.of("REMOVE '${stagingDirectory(streamConfig.id,
+            // database)}/${buffer.filename}'"))
+            workspaceClient
+                .files()
+                .delete("${stagingDirectory(streamConfig.id, database)}/${buffer.filename}")
+        }
+    }
+
+    fun typeAndDedupe(
+        streamConfig: StreamConfig,
+        maxProcessedTimestamp: Optional<Instant>,
+        finalTableSuffix: String
+    ) {
+        TypeAndDedupeTransaction.executeTypeAndDedupe(
+            sqlGenerator,
+            destinationHandler,
+            streamConfig,
+            maxProcessedTimestamp,
+            finalTableSuffix,
+        )
+    }
+
+    fun overwriteFinalTable(streamConfig: StreamConfig, suffix: String) {
+        // Guard to not accidentally overwrite existing table or DROP it.
+        if (suffix.isNotBlank()) {
+            log.info {
+                "Overwriting table ${streamConfig.id.finalTableId(DatabricksSqlGenerator.QUOTE)} with ${
+                    streamConfig.id.finalTableId(
+                        DatabricksSqlGenerator.QUOTE,
+                        suffix,
+                    )
+                }"
+            }
+            destinationHandler.execute(sqlGenerator.overwriteFinalTable(streamConfig.id, suffix))
+        }
+    }
+
+    fun prepareStagingTable(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
+        val rawSchema = streamId.rawNamespace
+        // TODO: Optimize by running SHOW SCHEMAS; rather than CREATE SCHEMA if not exists
+        destinationHandler.execute(sqlGenerator.createSchema(rawSchema))
+
+        // TODO: Optimize by running SHOW TABLES; truncate or create based on mode
+        // Create raw tables.
+        destinationHandler.execute(databricksSqlGenerator.createRawTable(streamId))
+        // Truncate the raw table if sync in OVERWRITE.
+        if (destinationSyncMode == DestinationSyncMode.OVERWRITE) {
+            destinationHandler.execute(databricksSqlGenerator.truncateRawTable(streamId))
+        }
+    }
+
+    fun createSchemaIfNotExists(streamConfig: StreamConfig) {
+        val finalSchema = streamConfig.id.finalNamespace
+        // TODO: Optimize by running SHOW SCHEMAS; rather than CREATE SCHEMA if not exists
+        destinationHandler.execute(sqlGenerator.createSchema(finalSchema))
+    }
+
+    fun createFinalTable(streamConfig: StreamConfig, suffix: String, replace: Boolean) {
```

Comment:
it would be cool to move away from a pure suffix strategy, since we know destinations have table name length limits

e.g. maybe this should be `createFinalTable(streamConfig: StreamConfig, nameOverride: String?, replace: Boolean)`? or something similar

and then we could have `nameOverride = streamConfig.id.finalName.substring(0, 32) + sha256(streamConfig.id.finalName + suffix).substring(0, 32)` or whatever.

Reply:
👌  Appreciate this. Forces us to think away from existing code boundaries